crates/abi/src/lib.rs

```diff
@@ -141,6 +141,65 @@ macro_rules! firmware_panic {
     };
 }
 
+// ————————————————————————————— Payload Setup —————————————————————————————— //
+
+#[macro_export]
+macro_rules! setup_payload {
+    ($path:path) => {
+        // The assembly entry point
+        core::arch::global_asm!(
+            r#"
+            .text
+            .align 4
+            .global _start
+            _start:
+                ld t4, __payload_bss_start
+                ld t5, __payload_bss_stop
+            zero_payload_bss_loop:
+                bgeu t4, t5, zero_payload_bss_done
+                sd x0, 0(t4)
+                addi t4, t4, 8
+                j zero_payload_bss_loop
+            zero_payload_bss_done:
+                // Load the stack pointer and jump into main
+                ld sp, __stack_top
+                j {entry}
+
+                // Store the address of the stack in memory
+                // That way it can be loaded as an absolute value
+            .align 8
+            __stack_top:
+                .dword {stack_top}
+            __payload_bss_start:
+                .dword {payload_bss_start}
+            __payload_bss_stop:
+                .dword {payload_bss_stop}
+            "#,
+            entry = sym _payload_start,
+            stack_top = sym _stack_top,
+            payload_bss_start = sym _payload_bss_start,
+            payload_bss_stop = sym _payload_bss_stop,
+        );
+
+        pub extern "C" fn _payload_start() -> ! {
+            // Validate the signature of the entry point.
+            let f: fn() -> ! = $path;
+
+            // Initialize logger
+            $crate::logger::init();
+
+            f();
+        }
+
+        // Defined in the linker script
+        extern "C" {
+            pub(crate) static _stack_top: u8;
+            pub(crate) static _payload_bss_start: u8;
+            pub(crate) static _payload_bss_stop: u8;
```

Comment:
I am wondering if we could just use the same as the firmware.

There are some minor differences, like the fact that we can't access `mhartid`, but we can work around. I'm not sure what is the best solution.

Reply:
I don't really know how to get over the `mhartid` read for the payload. Neither of the stack size and the `panic_handler` (if we want something different for our firmware and payloads).